health: Update Cilium agent to listen on nodeip

tamilmani1989 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->
This PR updates agent health server to listen on nodeip instead of all interfaces.
Fixes:
[#23353](https://github.com/cilium/cilium/issues/23353)

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

pippolo84 commented on cilium-health/responder/main.go:
```diff
@@ -31,7 +32,7 @@ func main() {
 	ctx, cancel := signal.NotifyContext(context.Background(), unix.SIGINT, unix.SIGHUP, unix.SIGTERM, unix.SIGQUIT)
 
-	srv := responder.NewServer(listen)
+	srv := responder.NewServer(node.GetCiliumEndpointNodeIP(), listen)
```
This is not going to work, because cilium-health-responder is a separate binary from the agent and here we don't start the `node` machinery that allows us to get the node IP:

```
# cilium-health-responder   
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x11c8687]

goroutine 1 [running]:
github.com/cilium/cilium/pkg/node.(*LocalNodeStore).Observe(0x16cf988?, {0x16cf950?, 0xc000426000?}, 0x13845a0?, 0x40d201?)
	<autogenerated>:1 +0x27
github.com/cilium/cilium/pkg/stream.First[...]({_, _}, {_, _})
	/go/src/github.com/cilium/cilium/pkg/stream/sinks.go:25 +0x1a3
github.com/cilium/cilium/pkg/node.(*LocalNodeStore).Get(...)
	/go/src/github.com/cilium/cilium/pkg/node/local_node_store.go:116
github.com/cilium/cilium/pkg/node.getLocalNode()
	/go/src/github.com/cilium/cilium/pkg/node/address.go:49 +0xad
github.com/cilium/cilium/pkg/node.GetIPv4()
	/go/src/github.com/cilium/cilium/pkg/node/address.go:282 +0x2e
github.com/cilium/cilium/pkg/node.GetCiliumEndpointNodeIP()
	/go/src/github.com/cilium/cilium/pkg/node/address.go:290 +0x29
main.main()
	/go/src/github.com/cilium/cilium/cilium-health/responder/main.go:35 +0x19a
```

Being a binary meant for testing purposes only, I think we can simply let it bind to all interfaces using an empty string as address.

tamilmani1989 replied:
my bad.. didnt realise `cilium-health-responder` separate binary.. let me fix this. Thanks for pointing out

pippolo84 reviewed (changes requested):
Thanks for your PR @tamilmani1989 . Changes look good on the agent side, but I think we should leave the cilium-health-responder behavior as is.

pippolo84 reviewed (commented):
Changes look good, thanks!

One last thing: could you please change the commit message to adhere to the guide [here](https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#submitting-a-pull-request)? Maybe a brief description of the solved issue and a reference to it in a `Fixes: ...` line. For the title, I suggest something like `health: Update daemon to listen on node ip`.
Thanks! :pray:

pippolo84 reviewed (approved):
Nice! :rocket:

pippolo84 commented on pkg/health/probe/responder/responder.go:
```diff
@@ -17,29 +18,62 @@ var defaultTimeout = 30 * time.Second
+		}
+
+		hs := http.Server{
+			Addr:    fmt.Sprintf("%s:%d", ip, port),
```
You can avoid this using [net.JoinHostPort](https://pkg.go.dev/net#JoinHostPort)

tamilmani1989 replied:
Thanks for pointing this out.

pippolo84 commented on pkg/health/probe/responder/responder.go:
```diff
@@ -17,29 +18,62 @@ var defaultTimeout = 30 * time.Second
-	return &Server{
-		http.Server{
-			Addr:    fmt.Sprintf(":%d", port),
+func NewServer(address []string, port int) *Server {
```
Since the code is now returning more servers, I think it'd be better to change the naming (e.g: `NewServer` -> `NewServers`, `httpServer` -> `httpServers` and so on), alongside their comments.

pippolo84 commented on pkg/health/probe/responder/responder.go:
```diff
@@ -17,29 +18,62 @@ var defaultTimeout = 30 * time.Second
+		}
+	}
+
+	return err
```
I suggest to join all the errors like this:

```go
	errs := make([]error, 0, len(s.httpServer))
	for _, hs := range s.httpServer {
		errs = append(errs, hs.Shutdown(ctx))
	}
	return errors.Join(errs...)
```
[errors,Join](https://pkg.go.dev/errors#Join) will discard any nil error, if any.

pippolo84 commented on pkg/health/probe/responder/responder_test.go:
```diff
@@ -0,0 +1,54 @@
+import (
+	"testing"
+
+	. "github.com/cilium/checkmate"
```
gopkg.in/check.v1 package was added to ease the transition out of `check.v1` in favor of the standard `testing` library. So, I suggest to use the std library for such a simple test (or, if you want to use assertion, you can try https://github.com/stretchr/testify that is already used widely in the codebase).

pippolo84 commented on pkg/health/server/server.go:
```diff
@@ -417,7 +419,24 @@ func NewServer(config Config) (*Server, error) {
+			address = append(address, nodeIpv6.String())
+		} else {
+			address = nil
+		}
```
I'd prefer to avoid these `else` and leave the slice as is, in case the ipv4 address has been already appended.
Maybe we should add a `log` entry in case `option.Config.EnableIPv4 == true && nodeIpv4 == nil` or `option.Config.EnableIPv6 == true && nodeIpv6 == nil`.

tamilmani1989 replied:
the reason I added this `else` is that if for some reason `GetInternalIPv6` returns nil, then health server should listen on all address (`:4240`)and not on just appended ipv4 otherwise `ipv6:port` would not be reachable. 

The other option is to listen on specific ipv4 address and listen on all ipv6 addr(`::`). The same should go for v4(`0.0.0.0`) if `GetInternalIPv4` returns nil

pippolo84 replied:
I'd go for the second option, seems more natural in terms of user expectations.

pippolo84 commented on pkg/node/types/node.go:
```diff
@@ -351,6 +351,21 @@ func (n *Node) GetK8sNodeIP() net.IP {
 }
 
+// GetNodeInternalIP returns the Internal IP of node or nil.
+func (n *Node) GetNodeInternalIP(ipv6 bool) net.IP {
```
The boolean flag in`GetNodeInternalIP(true)` and `GetNodeInternalIP(false)` is not so straightforward, I suggest to add two separate helpers (one for IPv4 and another one for IPv6) and get rid of it, so to improve the readability.
Also, that may simplify the `if` condition inside the helper itself.

tamilmani1989 replied:
sure will update it. I just followed existing api semantics `GetNodeIP` and `GetCiliumInternalIP` https://github.com/cilium/cilium/blob/6ba173d9ba392e05ff58930b482d697e00c70805/pkg/node/types/node.go#L356

pippolo84 reviewed (changes requested):
Thanks for this next iteration! :pray: 
I've left some comments, mainly around the style.

To setup a local cluster with external workloads you can refer to the doc [here](https://docs.cilium.io/en/latest/network/external-workloads/#external-workloads).
I think you can check `option.Config.JoinCluster` to understand if Cilium is running in a VM, since the feature is enabled based on that config flag.

Commit:
health: Update daemon to listen on node ip

Update agenthealth to listen on internal node ip instead of listening on
all interfaces except for external workload scenario where there is no
internal ip concept. Listen on ipv4 or ipv6 or both based on cilium
config EnableIPv4 and EnableIPv6. If for some reason, if it fails to get
node internal IP it will listen on all addresses.

Fixes: #23353

Signed-off-by: Tamilmani <tamanoha@microsoft.com>

pippolo84 commented on pkg/health/server/server.go:
```diff
@@ -417,7 +420,46 @@ func NewServer(config Config) (*Server, error) {
+	if option.Config.EnableIPv4 {
+		if ipv4 = node.GetInternalIPv4(); ipv4 != nil {
+			addresses = append(addresses, ipv4.String())
+		}
```
You can simply use a `else` here to account for the case `ipv4 == nil`, so to avoid the `if len(addresses) != 0` later.

pippolo84 commented on pkg/health/server/server.go:
```diff
@@ -417,7 +420,46 @@ func NewServer(config Config) (*Server, error) {
+	if option.Config.EnableIPv6 {
+		if ipv6 = node.GetInternalIPv6(); ipv6 != nil {
+			addresses = append(addresses, ipv6.String())
+		}
```
Same here for `ipv6`

pippolo84 reviewed (changes requested):
Thank you for the further iteration :100: 
I agree with your analysis regarding the `GetInternalIPv4` and `GetInternalIPv4` returnng `nil`. I've left a suggestion to improve the readability of the code when handling those cases.

pippolo84 reviewed (approved):
Thanks! This turned out to be trickier than expected, great job! :rocket:
